src/solvers/dg_simplices/dg.jl

```diff
@@ -0,0 +1,294 @@
+# !!! warning "Experimental features"
+
+# todo: replace mul! by Octavian.matmul!
+# todo: add @threaded 
+
+# out <- A*x
+mul_by!(A) = let A = A 
+  @inline (out, x)->mul!(out, A, x) 
+end
+
+# out <- out + A * x 
+mul_by_accum!(A) = let A = A 
+  @inline (out, x)->mul!(out, A, x, one(eltype(out)), one(eltype(out))) 
+end
+
+#  out <- out + α * A * x 
+mul_by_accum!(A, α) = let A = A 
+  @inline (out, x)->mul!(out, A, x, α, one(eltype(out))) 
+end
+
+const DGWeakForm{Dims, ElemType} = DG{<:RefElemData{Dims, ElemType}, Mortar, 
+                    <:SurfaceIntegralWeakForm,
+                    <:VolumeIntegralWeakForm} where {Mortar}
+
+# this is necessary for pretty printing
+Base.real(rd::RefElemData{Dims, Elem, ApproxType, Nfaces, RealT}) where {Dims, Elem, ApproxType, Nfaces, RealT} = RealT
+
+# iteration over quantities in a single element
+eachfacenode(mesh::AbstractMeshData, dg::DG{<:RefElemData}, cache) = Base.OneTo(dg.basis.Nfq)
+eachquadnode(mesh::AbstractMeshData, dg::DG{<:RefElemData}, cache) = Base.OneTo(dg.basis.Nq)
+
+# iteration over all elements in a mesh
+eachelement(mesh::AbstractMeshData, dg::DG{<:RefElemData}, cache) = Base.OneTo(mesh.md.num_elements)
+
+# iteration over quantities over the entire mesh (dofs, quad nodes, face nodes). 
+ndofs(mesh::AbstractMeshData, dg::DG{<:RefElemData}, cache) = dg.basis.Np * mesh.md.num_elements
+eachdoftotal(mesh::AbstractMeshData, dg::DG{<:RefElemData}, cache) = Base.OneTo(ndofs(mesh,dg,cache))
+eachquadnodetotal(mesh::AbstractMeshData, dg::DG{<:RefElemData}, cache) = Base.OneTo(dg.basis.Nq * mesh.md.num_elements)
+eachfacenodetotal(mesh::AbstractMeshData, dg::DG{<:RefElemData}, cache) = Base.OneTo(dg.basis.Nfq * mesh.md.num_elements)
+
+# interface with semidiscretization_hyperbolic
+wrap_array(u_ode::StructArray, mesh::AbstractMeshData, equations, dg::DG{<:RefElemData}, cache) = u_ode
+function digest_boundary_conditions(boundary_conditions::NamedTuple{Keys,ValueTypes}, mesh::AbstractMeshData, 
+                                    dg::DG{<:RefElemData}, cache) where {Keys,ValueTypes<:NTuple{N,Any}} where {N}
+  return boundary_conditions
+end
+
+function allocate_coefficients(mesh::AbstractMeshData, equations, dg::DG{<:RefElemData}, cache)
+  md = mesh.md
+  nvars = nvariables(equations) 
+  return StructArray{SVector{nvars, real(dg)}}(ntuple(_->similar(md.x),nvars))
+end
+
+function compute_coefficients!(u::StructArray, initial_condition, t,
+                        mesh::AbstractMeshData{Dim}, equations, dg::DG{<:RefElemData{Dim}}, cache) where {Dim}
+  md = mesh.md
+  rd = dg.basis
+  @unpack u_values = cache
+
+  for i in eachquadnodetotal(mesh,dg,cache)
```

Comment:
I have to admit that I have some trouble reading something like `eachquadnodetotal`, where four words are squashed together... Something like `eachnode` with only two words is fine for me, but I would prefer adding underscores or trying to find shorter names for `eachquadnodetotal`, if possible.

Reply:
I agree.

I can think of a few options:
1. Change the name: `each_quad_node_total` or `each_quad_node_global`
2. Add an outer loop over elements and use `eachnode` to iterate over nodes in an element.
3. We could use dispatch to distinguish between nodes on an element and nodes over the entire mesh (e.g., `each_quad_node(mesh, solver, cache)` for global nodes over the entire mesh, `each_quad_node(solver, cache)` for nodes over a single element.

I admit I don't like 2) since it would be easier to parallelize across all nodes with a single for loop. Do you have a preference between 1) or 3)

Reply:
I would prefer 1) since it's more clear to me.